server/src/main/java/org/elasticsearch/common/io/stream/RecyclerBytesStreamOutput.java

```diff
@@ -77,36 +94,90 @@ public void writeBytes(byte[] b, int offset, int length) {
 
         Objects.checkFromIndexSize(offset, length, b.length);
 
-        // get enough pages for new size
-        final int pageSize = this.pageSize;
-        int currentPageOffset = this.currentPageOffset;
+        final int currentPageOffset = this.currentPageOffset;
         if (length > pageSize - currentPageOffset) {
-            ensureCapacity(length);
-            currentPageOffset = this.currentPageOffset;
+            writeMultiplePages(b, offset, length);
+        } else {
```

Comment:
Specifically add a branch for the extremely common scenario where the bytes fit in the current page.

Reply:
I actually pulled this change out as I don't think the evidence supports it yet.